File: eprapiservices/src/main/java/epr/eprapiservices/service/MaterialService.java
package epr.eprapiservices.service;

import epr.eprapiservices.entity.Material;
import epr.eprapiservices.dao.repository.MaterialRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class MaterialService {

    @Autowired
    private MaterialRepository materialRepository;

    /**
     * Get all active materials with their material type information
     */
    @Transactional(readOnly = true)
    public List<Material> getAllMaterials() {
        return materialRepository.findAllActiveMaterialsWithType();
    }

    /**
     * Get material by ID with material type information
     */
    @Transactional(readOnly = true)
    public Optional<Material> getMaterialById(Integer id) {
        return materialRepository.findByIdWithType(id);
    }

    /**
     * Get material by material code
     */
    @Transactional(readOnly = true)
    public Optional<Material> getMaterialByCode(String materialCode) {
        return materialRepository.findByMaterialCode(materialCode);
    }

    /**
     * Get materials by active status
     */
    @Transactional(readOnly = true)
    public List<Material> getMaterialsByActiveStatus(Boolean isActive) {
        return materialRepository.findByIsActive(isActive);
    }

    /**
     * Search materials by name
     */
    @Transactional(readOnly = true)
    public List<Material> searchMaterialsByName(String searchTerm) {
        return materialRepository.findByMaterialNameContainingIgnoreCase(searchTerm);
    }

    /**
     * Create a new material
     */
    public Material createMaterial(Material material) {
        // Check if material code already exists
        if (materialRepository.existsByMaterialCodeAndNotMaterialId(material.getMaterialCode(), null)) {
            throw new IllegalArgumentException("Material code '" + material.getMaterialCode() + "' already exists");
        }

        // Set default values
        if (material.getSortOrder() == null) {
            material.setSortOrder(1);
        }
        material.setIsActive(true);

        return materialRepository.save(material);
    }

    /**
     * Update an existing material
     */
    public Material updateMaterial(Integer id, Material materialDetails) {
        Material existingMaterial = materialRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Material with ID " + id + " not found"));

        // Check if material code already exists (excluding current material)
        if (materialRepository.existsByMaterialCodeAndNotMaterialId(materialDetails.getMaterialCode(), id)) {
            throw new IllegalArgumentException("Material code '" + materialDetails.getMaterialCode() + "' already exists");
        }

        // Update fields
        existingMaterial.setMaterialName(materialDetails.getMaterialName());
        existingMaterial.setMaterialCode(materialDetails.getMaterialCode());
        existingMaterial.setDescription(materialDetails.getDescription());
        existingMaterial.setSortOrder(materialDetails.getSortOrder());
        // BaseModel will handle updatedDate automatically via @PreUpdate

        return materialRepository.save(existingMaterial);
    }

    /**
     * Hard delete a material (permanently remove from database)
     */
    public void deleteMaterial(Integer id) {
        if (!materialRepository.existsById(id)) {
            throw new IllegalArgumentException("Material with ID " + id + " not found");
        }

        materialRepository.deleteById(id);
    }

    /**
     * Soft delete a material (set isActive to false)
     */
    public void softDeleteMaterial(Integer id) {
        Material material = materialRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Material with ID " + id + " not found"));

        material.setIsActive(false);
        // BaseModel will handle updatedDate automatically via @PreUpdate
        materialRepository.save(material);
    }

    /**
     * Check if material exists
     */
    @Transactional(readOnly = true)
    public boolean existsById(Integer id) {
        return materialRepository.existsById(id);
    }

    /**
     * Count all active materials
     */
    @Transactional(readOnly = true)
    public long countActiveMaterials() {
        return materialRepository.findByIsActive(true).size();
    }
}

File: eprwebui/src/services/materialService.ts
interface Material {
  materialId?: number;
  materialCode: string;
  materialName: string;
  description?: string;
  sortOrder?: number;
  isActive?: boolean;
  createdDate?: string;
  updatedDate?: string;
}

class MaterialService {
  private baseURL = 'http://localhost:8080/api/materials';

  async getAllMaterials(): Promise<Material[]> {
    const response = await fetch(this.baseURL);
    if (!response.ok) {
      throw new Error('Failed to fetch materials');
    }
    return response.json();
  }

  async getMaterialById(id: number): Promise<Material> {
    const response = await fetch(`${this.baseURL}/${id}`);
    if (!response.ok) {
      throw new Error('Failed to fetch material');
    }
    return response.json();
  }

  async createMaterial(material: Material): Promise<Material> {
    const response = await fetch(this.baseURL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(material),
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(errorText || 'Failed to create material');
    }
    
    return response.json();
  }

  async updateMaterial(id: number, material: Material): Promise<Material> {
    const response = await fetch(`${this.baseURL}/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(material),
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(errorText || 'Failed to update material');
    }
    
    return response.json();
  }

  async deleteMaterial(id: number): Promise<void> {
    const response = await fetch(`${this.baseURL}/${id}`, {
      method: 'DELETE',
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(errorText || 'Failed to delete material');
    }
  }

  async searchMaterials(query: string): Promise<Material[]> {
    const response = await fetch(`${this.baseURL}/search?query=${encodeURIComponent(query)}`);
    if (!response.ok) {
      throw new Error('Failed to search materials');
    }
    return response.json();
  }

  async getMaterialsByActiveStatus(isActive: boolean): Promise<Material[]> {
    const response = await fetch(`${this.baseURL}/active/${isActive}`);
    if (!response.ok) {
      throw new Error('Failed to fetch materials by status');
    }
    return response.json();
  }

  async countActiveMaterials(): Promise<number> {
    const response = await fetch(`${this.baseURL}/count/active`);
    if (!response.ok) {
      throw new Error('Failed to count active materials');
    }
    return response.json();
  }
}

export default new MaterialService();
